marker_api/src/sem/ty.rs

```diff
@@ -110,3 +114,43 @@ impl<'ast> TyKind<'ast> {
         ty
     }
 }
+
+#[repr(C)]
+#[cfg_attr(feature = "driver-api", visibility::make(pub))]
+#[cfg_attr(feature = "driver-api", derive(typed_builder::TypedBuilder))]
+pub(crate) struct CommonTyData<'ast> {
+    #[cfg_attr(feature = "driver-api", builder(default))]
+    _lifetime: PhantomData<&'ast ()>,
+    driver_id: DriverTyId,
+}
+
+impl<'ast> Debug for CommonTyData<'ast> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CommonTyData {...}").finish()
+    }
+}
+
+#[cfg(feature = "driver-api")]
+impl<'ast> CommonTyData<'ast> {
+    pub fn driver_id(&self) -> DriverTyId {
+        self.driver_id
+    }
+}
+
+macro_rules! impl_ty_data {
+    ($self_ty:ty, $enum_name:ident) => {
+        #[cfg(feature = "driver_api")]
+        impl<'ast> $self_ty {
+            pub fn data(&self) -> &$crate::sem::ty::CommonTyData<'ast> {
```

Comment:
I'd expect this to fail to compile because this is a `pub` function which returns a `pub(crate)` type.